core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java

```diff
@@ -312,25 +312,30 @@ public int hashCode() {
         return nodeId.hashCode();
     }
 
+    private String toString;
```

Comment:
non volatile multi thread assignment/read - dangerous territory , but we're good here since we only over calculate in this case  (working with the JVM will be crazy if would expose half baked objects in this case).